File: linked-list-cycle-ii/linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if (!head || !head->next) return NULL; 
        ListNode* fast = head;
        ListNode* slow = head;
        while (fast != NULL && fast->next != NULL) {
            slow = slow->next;
            fast = fast->next;
            if (fast) fast = fast->next;
            if (fast == slow) {
                while(slow!=head)  // to find where the loop  started
            {
                slow=slow->next;
                head=head->next;
            }
            return head;
            }
        }
        return NULL;
    }
};

File: linked-list-cycle-ii/linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next: return
        
        fast = head
        slow = head
        
        while fast and fast.next:
            fast = fast.next
            slow = slow.next
            if fast: fast = fast.next
            if fast == slow:
                while slow != head:
                    slow = slow.next
                    head = head.next
                return head
        
        return
        